### ISSUE TITLE
Multiband Compressor: Phase issue in "Classic" mode

### ISSUE THREAD

**Issue by dstra2** [sadko4u/lsp-plugins#210]
Hi,

I was trying to locate the cause of Multiband Compressor's phase issue in "Classic" mode and found the following piece of code for the calculation of the phase compensation allpass filter (Filter.cpp, lines 1326-1370, function Filter::calc_lrx_filter):

```
case FLT_BT_LRX_ALLPASS:
            {
                float k     = 1.0f / (1.0f + fp->fQuality);
                size_t i    = sParams.nSlope * 2;

                // Emit 2x butterworth filters
                for (size_t j=0; j < i; j += 2)
                {
                    float theta     = ((j+1) * M_PI_2)/i;
                    float tsin      = sinf(theta);
                    float tcos      = sqrtf(1.0 - tsin*tsin);
                    float kf        = tsin*tsin + k*k * tcos*tcos;

                    c1              = add_cascade();
                    c2              = add_cascade();

                    // Top part
                    float xeta      = ((j+0.5) * M_PI)/i;
                    c1->t[0]        = 1.0;
                    c1->t[1]        = -2.0 * cosf(xeta);
                    c1->t[2]        = 1.0;

                    xeta            = ((j+1.5) * M_PI)/i;
                    c2->t[0]        = 1.0;
                    c2->t[1]        = -2.0 * cosf(xeta);
                    c2->t[2]        = 1.0;

                    // Bottom part
                    c1->b[0]        = 1.0 / kf;
                    c1->b[1]        = 2.0 * k * tcos / kf;
                    c1->b[2]        = 1.0;

                    c2->b[0]        = c1->b[0];
                    c2->b[1]        = c1->b[1];
                    c2->b[2]        = c1->b[2];

                    if (j == 0)
                    {
                        c1->t[0]       *= fp->fGain;
                        c1->t[1]       *= fp->fGain;
                        c1->t[2]       *= fp->fGain;
                    }
                }
                return;
            }
```
Although I do not fully understand the calculation of the coefficients, the code seems to create a cascade of **two** allpass biquads. But as far as I understand the Linkwitz-Riley crossver, only **one** allpass biquad is required to compensate for the phase shift caused by the two cascaded low- or highpass biquads of the crossover (LR4) because the phase shift of a single allpass biquad is exactly twice as large as the phase shift of the corresponding low- or highpass biquad:
### Lowpass
![lowpass](https://user-images.githubusercontent.com/24780672/141380675-b80f1aab-18a5-479d-92f5-5fe945a75a8e.png)
### Highpass
![highpass](https://user-images.githubusercontent.com/24780672/141380686-ff688840-deae-4c39-a649-1107938bf9e2.png)
### Allpass
![allpass](https://user-images.githubusercontent.com/24780672/141380692-0a6376d7-254f-4d8b-9997-d957a72a8c54.png)
(Source: [https://www.earlevel.com/main/2021/09/02/biquad-calculator-v3/](https://www.earlevel.com/main/2021/09/02/biquad-calculator-v3/))

What do you think, could this be the root cause of the phase issue?

**Comment by sadko4u** [comment]
> But as far as I understand the Linkwitz-Riley crossver, only one allpass biquad is required to compensate for the phase shift caused by the two cascaded low- or highpass biquads of the crossover (LR4) because the phase shift of a single allpass biquad is exactly twice as large as the phase shift of the corresponding low- or highpass biquad

The understanding is not completely correct. That's true for first (or second) order Linkwitz-Riley filters.
But when we use high-order Linkwitz-Riley filters, the phase compensation is not enough.

Instead of this, the allpass filter should be computed in the same way as the signal was processed by the lowpass filter, hipass filter, and the processed signals have been summed together. So if we have transfer functions:
- Hlpf(p) = 1 / (1 - p) - for low-pass filter
- Hhpf(p) = p / (1 - p) - for high-pass filter
the resulting all-pass filter should look like:
- Hapf(p) = Hlpf(p) + Hhpf(p) =  1 / (1 - p) + p / (1 - p) = (1 + p) / (1 - p)

In the same way the transfer function can be computed for high-order low-pass, high-pass and all-pass filters.

**Comment by sadko4u** [comment]
And, additionally, what's the phasing issue with Multiband Compressor you're observing?

**Comment by dstra2** [comment]
Thank you for the explanation about the allpass filter (and for posing the right question...). I did some tests and it seems that the phase issue occurs only when applying parallel compression, i.e. when the dry signal is mixed with the wet signal (within the compressor).

![Parallel Compression](https://user-images.githubusercontent.com/24780672/141660738-0b7be145-9664-458b-ae1a-84ab1c15c234.png)

I therefore can confirm that the cause cannot lie in the filter calculation. I rather suppose that in the dry signal path no phase compensation is taking place in "Classic" mode.

**Comment by sadko4u** [comment]
Yes, there are no changes for the 'dry' signal.

**Comment by sadko4u** [comment]
But it is possible to provide some option like 'Compensate phase for dry signal' or add parallel compression facility for each band.

**Comment by dstra2** [comment]
I would suggest to compensate the phase of the dry signal instead of adding a parallel compression facility for each band. There are two reasons for this:

1. I am not aware of any use case that requires parallel compression per band. If there are such use cases, they can also be realised by upward compression (per band), since the compression curve of the upward compressor ("Boosting" mode with large knee) is very similar to that of parallel compression.

2. By using the following structure, the phase compensation can be realised computationally efficiently by only 7 additional additions per sample (= number of bands - 1) and without additional multiplications.
![Multiband Compressor](https://user-images.githubusercontent.com/24780672/142504810-350b1b15-efca-4e34-8365-af73d167ff48.gif)

I would also suggest enabling phase compensation automatically as soon as the "Dry" volume control is set to greater than -inf dB instead of adding an additional button in the GUI.

What do you think?

**Comment by sadko4u** [comment]
1. It seems to be reasonable.
2. I think there still should be a button for `Dry` signal that enables the phase compensation but this button should be enabled by default. It may be a case that someone can find the phase distortion the exact thing for the project.
3. Also, these changes should affect other multiband plugins.

Since we're currently very busy by bringing 1.2.0 to life, the changes won't be earler than 1.2.0 comes out.

**Comment by sadko4u** [comment]
Merged changes for all multiband plugins:
https://github.com/lsp-plugins/lsp-plugins-mb-compressor/commit/571ccbdad6ba47214b467ec2675d11c5db4b1c12
https://github.com/lsp-plugins/lsp-plugins-mb-expander/commit/cdab9f7d297b07986346240d939c9ea4b6112f79
https://github.com/lsp-plugins/lsp-plugins-mb-gate/commit/c2aeb9f7c72fbb6ae45123738195ae05b3b2211b

**Comment by sadko4u** [comment]
Available in 1.2.1.
